Rac dropzone/filetrigger docs (#4847)
adobe/react-spectrum: packages/react-aria-components/docs/FileTrigger.mdx

* **Customizable** – Works with any pressable React Aria or React Spectrum component, and custom components built with [usePress](usePress.html).

## Anatomy

A `FileTrigger` wraps around a pressable child such as a button, and includes a visually hidden input element that allows the user to select files from their device.

If a visual label is not provided on the pressable child, then an `aria-label` or `aria-labelledby` prop must be passed to identify the file trigger to assistive technology.

### Composed Components
A `FileTrigger` can use the following components, which may also be used standalone or reused in other components.

<ExampleCard
  url="Button.html"
  title="Button"
  description="A button allows a user to perform an action with a mouse, touch, or keyboard.">
  <Button />
</ExampleCard>

## Props

<PropTable component={docs.exports.FileTrigger} links={docs.links} />

## Styling

> React Aria components can be styled in many ways, including using CSS classes, inline styles, utility classes (e.g. Tailwind), CSS-in-JS (e.g. Styled Components), etc. Unlike other React Aria components, `FileTrigger` does not have a builtin `className` attribute following the `react-aria-ComponentName` naming convention. Instead, it is styled by its pressable children, such as `Button`.  

Reviewer: Wondering if we should just remove the styling section completely for this one since styling a button is covered on the button page? Or change it to say something like we have for DialogTrigger/TooltipTrigger? https://reactspectrum.blob.core.windows.net/reactspectrum/f167e7f31c3e00033cf52520ce52f3de4d8d5ccf/docs/react-aria/Tooltip.html#tooltiptrigger-1 The first sentence here isn't really relevant, and we could just say that since it doesn't render any elements of its own it doesn't support styling.

the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import {Layout} from '@react-spectrum/docs';
export default Layout;

import docs from 'docs:react-aria-components';
import typesDocs from 'docs:@react-types/shared/src/events.d.ts';
import {PropTable, HeaderInfo, TypeLink, PageDescription, StateTable} from '@react-spectrum/docs';
import styles from '@react-spectrum/docs/src/docs.css';
import packageData from 'react-aria-components/package.json';
import Anatomy from '@react-aria/datepicker/docs/daterangepicker-anatomy.svg';
import ChevronRight from '@spectrum-icons/workflow/ChevronRight';
import {Divider} from '@react-spectrum/divider';
import {ExampleCard} from '@react-spectrum/docs/src/ExampleCard';
import {Keyboard} from '@react-spectrum/text';
import Button from '@react-spectrum/docs/pages/assets/component-illustrations/ActionButton.svg';
import Link from '@react-spectrum/docs/pages/assets/component-illustrations/Link.svg';

---

> category: Drag and drop

Reviewer: Topic for discussion with the team on this category

  --text-color-disabled: var(--spectrum-alias-text-color-disabled);
  --focus-ring-color: slateblue;

  color: var(--text-color);
  background: var(--background-color);
  border: 1px solid var(--border-color);
  border-radius: 4px;
  appearance: none;
  vertical-align: middle;
  font-size: 1.2rem;
  text-align: center;
  margin: 0;
  outline: none;
  padding: 4px 12px;

  &[data-pressed] {
    box-shadow: inset 0 1px 2px rgb(0 0 0 / 0.1);
    background: var(--background-color-pressed);
    border-color: var(--border-color-pressed);
  }

  &[data-focus-visible] {
    border-color: var(--focus-ring-color);
    box-shadow: 0 0 0 1px var(--focus-ring-color);
  }

  &:disabled {
    border-color: var(--border-color-disabled);
    color: var(--text-color-disabled);
  }
}

@media (forced-colors: active) {
  .react-aria-Button {
    forced-color-adjust: none;
    --border-color: ButtonBorder;
    --border-color-pressed: ButtonBorder;
    --border-color-disabled: GrayText;
    --background-color: ButtonFace;
    --background-color-pressed: ButtonFace;
    --text-color: ButtonText;
    --text-color-disabled: GrayText;
    --focus-ring-color: Highlight;
  }
}
```

</details>

## Features

> Access to a computer’s file system can be created in HTML using an `<input>` with a file `type` attribute. However, file input's on their own are not very useful since they only allow users to select files from their devices. As a result, file inputs are commonly used inside a `<form>` alongside  a `<button>` to submit the form. `FileTrigger` simplifies this by setting up the input element's configuration with reference to a pressable component, such as a button, for a customizable file upload. 

Reviewer: I think we should focus this on allowing custom styling. FileTrigger is basically just a hidden `<input type="file">` that is triggered by a custom button. Maybe this could be a starting point:

"A file input can be created with an `<input type="file">` element, but this supports very limited styling options. `FileTrigger` works with a pressable child such as a `Button` to create accessible file inputs that can be styled as needed."

> * **Customizable** – Works with any pressable component meaning that it is compatible with pressable React Aria Components and React Spectrum Components.

Reviewer: ```suggestion
* **Customizable** – Works with any pressable React Aria or React Spectrum component, and custom components built with [usePress](usePress.html).
```

## Anatomy

> A `FileTrigger` wraps around an input element and a pressable child, usually a button or link, that allows the user to upload files. The input element is not displayed and has no effect on the layout. 

Reviewer: ```suggestion
A `FileTrigger` wraps around a pressable child such as a button, and includes a visually hidden input element that allows the user to select files from their device.
```

> If a visual label is not provided on the pressable child, then an `aria-label` or `aria-labelledby` prop must be passed to identify the link to assistive technology.

Reviewer: ```suggestion
If a visual label is not provided on the pressable child, then an `aria-label` or `aria-labelledby` prop must be passed to identify the file trigger to assistive technology.
```

### Composed Components
A `FileTrigger` can use the following components, which may also be used standalone or reused in other components.

<ExampleCard
  url="Button.html"
  title="Button"
  description="A button allows a user to perform an action with a mouse, touch, or keyboard.">
  <Button />
</ExampleCard>

> <ExampleCard

Reviewer: I'd like to discourage using links as file triggers. It doesn't semantically make much sense since a link is meant to navigate a user to another page, and this is not doing that. I think we should just focus our examples on buttons for now. If someone wants to make their design visually look like a link, they can style a button to just look like underlined text.

  url="Link.html"
  title="Link"
  description="A link allows a user to navigate to another page.">
  <Link />
</ExampleCard>

## Props

> <PropTable component={docs.exports.FileTrigger} links={docs.links} />

Reviewer: Looks like the aria labeling props are on here but not used. I think we can remove these from the TypeScript definition since they should just be added to the `<Button>` instead.

We should probably also add the `name` prop here and pass it to the `<input>` element so that the selected file can be submitted as part of a form.

## Styling

React Aria components can be styled in many ways, including using CSS classes, inline styles, utility classes (e.g. Tailwind), CSS-in-JS (e.g. Styled Components), etc. By default, all components include a builtin `className` attribute which can be targeted using CSS selectors. These follow the `react-aria-ComponentName` naming convention.

```css

> .react-aria-FileTrigger {

Reviewer: Now thinking about this, I'm wondering if we should actually remove the extra wrapper div around FileTrigger, since it isn't really needed (you'd usually just style the Button inside). Both DialogTrigger and TooltipTrigger don't have an extra div, so perhaps making FileTrigger match that would be a good idea.

If we decide to do that, then we can also remove `SlotProps` and `DOMProps` from the type definition.

  /* ... */
}
```

A custom `className` can also be specified on any component. This overrides the default `className` provided by React Aria with your own.

```jsx
<FileTrigger className="my-filetrigger">
  {/* ... */}
</FileTrigger>
```

The states and selectors for each component used in a `FileTrigger` are documented below.

### Button

A [Button](Button.html) can be targeted with the `.react-aria-Button` CSS selector, or by overriding with a custom `className`. It supports the following states:

<StateTable properties={docs.exports.ButtonRenderProps.properties} />

> ### Link

Reviewer: same as above, I think we can remove this section for now
